linkerd/http-route/src/http/filter/inject_failure.rs

```diff
@@ -0,0 +1,74 @@
+use std::hash::Hash;
+
+/// A filter that responds with an HTTP error.
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct InjectFailure {
+    pub response: FailureResponse,
+    pub distribution: Distribution,
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct FailureResponse {
+    pub status: http::StatusCode,
+    pub message: std::sync::Arc<str>,
+}
+
+/// A Bernoulli distribution that implements `Hash`, `PartialEq`, and `Eq`.
+#[derive(Clone, Debug)]
+pub struct Distribution {
+    numerator: u32,
+    denominator: u32,
+    inner: rand::distributions::Bernoulli,
+}
+
+// === impl InjectFailure ===
+
+impl InjectFailure {
+    pub fn apply(&self) -> Option<FailureResponse> {
+        use rand::distributions::Distribution;
+
+        if self.distribution.sample(&mut rand::thread_rng()) {
+            return Some(self.response.clone());
+        }
+
+        None
+    }
+}
+
+// === impl InjectFailure ===
+
+impl Distribution {
+    pub fn from_ratio(
+        numerator: u32,
+        denominator: u32,
+    ) -> Result<Self, rand::distributions::BernoulliError> {
+        let inner = rand::distributions::Bernoulli::from_ratio(numerator, denominator)?;
```

Comment:
this seems like it could eventually be a cool use case for the [`fast_bernoulli` crate](https://crates.io/crates/fast-bernoulli), rather than `rand` --- that crate avoids generating random numbers when an event is *not* sampled, which could be a kind of nice optimization. but, we can always play with that later; it's not a blocker for now!

Reply:
yeah, if we find other uses for that crate, i'd be more inclined to use it here. it doesn't seem worth a dependency just for this use case though.